### PR TITLE
[Backport] Wallet database handling abstractions/simplifications

### PR DESCRIPTION
**Pull request by furszy**
Back ported #9951 + cleaned all of the zerocoin related, not used, methods inside the walletdb class.

Built on top of #2082.

### CODE REVIEW

**Comment by random-zebra** [comment]
Can be rebased (#2082 merged)

**Comment by furszy** [comment]
Rebased on top of master. Ready for review.

File: src/wallet/walletdb.cpp [hunk 1]
```diff
@@ -957,18 +957,18 @@ bool BackupWallet(const CWallet& wallet, const fs::path& strDest, bool fEnableCu
...
             LOCK(bitdb.cs_db);
-            if (!bitdb.mapFileUseCount.count(wallet.strWalletFile) || bitdb.mapFileUseCount[wallet.strWalletFile] == 0) {
+            if (!bitdb.mapFileUseCount.count(wallet.GetDBHandle().GetName()) || bitdb.mapFileUseCount[wallet.GetDBHandle().GetName()] == 0) {
                 // Flush log data to the dat file
-                bitdb.CloseDb(wallet.strWalletFile);
-                bitdb.CheckpointLSN(wallet.strWalletFile);
-                bitdb.mapFileUseCount.erase(wallet.strWalletFile);
+                bitdb.CloseDb(wallet.GetDBHandle().GetName());
+                bitdb.CheckpointLSN(wallet.GetDBHandle().GetName());
+                bitdb.mapFileUseCount.erase(wallet.GetDBHandle().GetName());
```
[the remark this answers is not in the document]
Comment (random-zebra) [reply src/wallet/walletdb.cpp:960]: Actually we should just make use of `CWalletDBWrapper::Backup` here

**Review by random-zebra — DISMISSED** [review]
Looking good. Just few nits.
The `CWalletDBWrapper::Backup` function, introduced here, is currently never used. 
We should refactor `BackupWallet` (which duplicates most of the code), moving it to CWallet, and making it just a wrapper around this new `CWalletDBWrapper::Backup` (with proper handling of the custom path).

In the future, we should probably also better encapsulate the methods that deal with the DB, inside `CWallet`, to avoid using `GetDBHandle` in so many places (e.g., the RPC, the qt options, or the spkm) which shouldn't have knowledge of lower level db.

**Comment by furszy** [comment]
> The CWalletDBWrapper::Backup function, introduced here, is currently never used.
We should refactor BackupWallet (which duplicates most of the code), moving it to CWallet, and making it just a wrapper around this new CWalletDBWrapper::Backup (with proper handling of the custom path).

Yes, absolutely. Was planning to do it on another PR but can tackle it here as well.

> In the future, we should probably also better encapsulate the methods that deal with the DB, inside CWallet, to avoid using GetDBHandle in so many places (e.g., the RPC, the qt options, or the spkm) which shouldn't have knowledge of lower level db.

Yep, had the same thoughts but it was too much to wrap it here as well.

**Comment by furszy** [comment]
PR updated, decoupled the custom backup wallet path and the historical backup code block from the static `BackupWallet` into their own functions.

This is a previous step before be able to combine and generalize the two backup functions that we currently have into a single one. Will work on it in another PR as the historical backups code is tangled with the GUI notification. Plus.. really needs a focused review.

File: src/wallet/walletdb.cpp [hunk 2]
```diff
@@ -928,27 +928,35 @@ void NotifyBacked(const CWallet& wallet, bool fSuccess, std::string strMessage)
...
+}
+
+bool BackupWallet(const CWallet& wallet, const fs::path& strDest)
+{
+    Optional<fs::path> opPathCustom = GetBackupPath(wallet);
+    if (!opPathCustom) return false;
+
+    // Custom backup dir and subsequent dir/file.dat
+    fs::path pathCustom = *opPathCustom;
+    fs::path pathWithFile = pathCustom / wallet.GetUniqueWalletBackupName();
```
Comment (random-zebra) [inline src/wallet/walletdb.cpp:1014]: We shouldn't return early when no custom path is provided, we should continue with empty `pathWithFile`.
I think we can remove the Optional and just return the (possibly empty) `fs::path` from `GetBackupPath`.

Comment (furszy) [reply src/wallet/walletdb.cpp:1014]: good catch 👍 .

**Comment by furszy** [comment]
Updated per @random-zebra's feedback.

**Review by random-zebra — APPROVED** [review]
utACK 745e6fccbfeb1c2d1e544941366c7f2b73d14ea1

Would be good to add a test for the backupwallet RPC.
We can add it in the follow-up PR that will refactor `BackWallet` using `CWalletDBWrapper::Backup`.

File: src/wallet/walletdb.cpp [hunk 3]
```diff
@@ -1047,15 +1055,8 @@ bool AttemptBackupWallet(const CWallet& wallet, const fs::path& pathSrc, const f
         fs::copy_file(pathSrc.c_str(), pathDest, fs::copy_options::overwrite_existing);
 #elif BOOST_VERSION >= 105800 /* BOOST_LIB_VERSION 1_58 */
         fs::copy_file(pathSrc.c_str(), pathDest, fs::copy_option::overwrite_if_exists);
-#else
-        std::ifstream src(pathSrc.c_str(),  std::ios::binary | std::ios::in);
-        std::ofstream dst(pathDest.c_str(), std::ios::binary | std::ios::out | std::ios::trunc);
-        dst << src.rdbuf();
-        dst.flush();
-        src.close();
-        dst.close();
```
Comment (Fuzzbawls) [inline src/wallet/walletdb.cpp:1056]: Removing this fallback block breaks compatibility with Boost 1.57, which is our current minimum supported version. so it should probably be left in place until we decide to bump the minimum boost version.

**Review by Fuzzbawls — COMMENTED** [review]
Overall good, but introduced a compatibility issue with our current minimum supported version of Boost

File: src/wallet/walletdb.cpp (hunk 3 above)
Comment (furszy) [reply src/wallet/walletdb.cpp:1056]: done

**Review by random-zebra — APPROVED** [review]
re-utACK e5fd21578df210e96730bbf22f99fb96e8752223

**Review by Fuzzbawls — APPROVED** [review]
utACK e5fd21578df210e96730bbf22f99fb96e8752223

**Pull request merged**